core/src/main/java/org/pentaho/di/connections/vfs/provider/ConnectionFileObject.java

```diff
@@ -126,132 +168,163 @@ public ConnectionFileObject( AbstractFileName name, ConnectionFileSystem fs,
+    return resolvedFileObject != null && resolvedFileObject.isExecutable();
   }
 
-  @Override public boolean isFile() throws FileSystemException {
-    return resolvedFileObject.isFile();
+  @Override
+  public boolean isFile() throws FileSystemException {
+    return resolvedFileObject != null && resolvedFileObject.isFile();
   }
 
-  @Override public boolean isFolder() throws FileSystemException {
-    return resolvedFileObject.isFolder();
+  @Override
+  public boolean isFolder() throws FileSystemException {
+    return resolvedFileObject != null ? resolvedFileObject.isFolder() : isPvfsRoot();
   }
 
-  @Override public boolean isHidden() throws FileSystemException {
-    return resolvedFileObject.isHidden();
+  @Override
+  public boolean isHidden() throws FileSystemException {
+    return resolvedFileObject != null && resolvedFileObject.isHidden();
   }
 
-  @Override public boolean isReadable() throws FileSystemException {
-    return resolvedFileObject.isReadable();
+  @Override
+  public boolean isReadable() throws FileSystemException {
+    return resolvedFileObject != null && resolvedFileObject.isReadable();
   }
 
-  @Override public boolean isWriteable() throws FileSystemException {
-    return resolvedFileObject.isWriteable();
+  @Override
+  public boolean isWriteable() throws FileSystemException {
+    // TODO: should let the base class handle it and then implement doWriteable instead.
+    return ensureResolvedFileObject().isWriteable();
   }
 
-  @Override public void moveTo( FileObject destFile ) throws FileSystemException {
-    resolvedFileObject.moveTo( destFile );
+  @Override
+  public void moveTo( FileObject destFile ) throws FileSystemException {
+    ensureResolvedFileObject().moveTo( destFile );
   }
 
-  @Override public void refresh() throws FileSystemException {
+  @Override
+  public void refresh() throws FileSystemException {
     if ( resolvedFileObject != null ) {
       resolvedFileObject.refresh();
     }
   }
 
+  @Nullable
   public FileObject getResolvedFileObject() {
     return resolvedFileObject;
   }
 
+  @NonNull
+  protected AbstractFileObject<?> ensureResolvedFileObject() throws FileSystemException {
+    if ( resolvedFileObject == null ) {
+      if ( isPvfsRoot() ) {
+        throw new UnsupportedOperationException( "Operation not supported or implemented at the PVFS root level." );
+      }
+
+      throw new FileSystemException( "vfs.provider/connection-or-file-not-found", getName().toString() );
+    }
+
+    return resolvedFileObject;
+  }
+
   @Override
   public String getOriginalURIString() {
     return this.getName().toString();
   }
 
   @Override
   public String getAELSafeURIString() {
-    return resolvedFileObject.getPublicURIString().replaceFirst( "s3://", "s3a://" );
+    try {
+      return ensureResolvedFileObject().getPublicURIString().replaceFirst( "s3://", "s3a://" );
+    } catch ( FileSystemException e ) {
+      throw new RuntimeException( e );
```

Comment:
Trying to preserve backwards compatibility in a method that does not have checked exceptions...